File: isegoria/src/main/java/com/eulersbridge/isegoria/feed/FeedFragment.java
package com.eulersbridge.isegoria.feed;

import android.content.Context;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.eulersbridge.isegoria.MainActivity;
import com.eulersbridge.isegoria.util.Utils;
import com.eulersbridge.isegoria.util.ui.TitledFragment;
import com.eulersbridge.isegoria.R;


public class FeedFragment extends Fragment implements TitledFragment, MainActivity.TabbedFragment {

    private TabLayout tabLayout;
    private ViewPager viewPager;

    @Override
	public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.feed_fragment, container, false);

        AppCompatActivity activity = (AppCompatActivity)getActivity();

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP && activity != null) {
            int colour = ContextCompat.getColor(activity, R.color.darkBlue);
            activity.getWindow().setStatusBarColor(colour);

            Utils.setMultitaskColour(activity, colour);
            Utils.setStatusBarColour(activity, colour);
        }

        if (activity != null) {
            // Ensure options menu from another fragment is not carried over
            activity.invalidateOptionsMenu();

            Utils.hideKeyboard(activity);
        }

        setHasOptionsMenu(true);

        setupViewPager(rootView);

		return rootView;
	}

	public String getTitle(Context context) {
	    return context.getString(R.string.section_title_feed);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.feed, menu);
        super.onCreateOptionsMenu(menu, inflater);

        MenuItem friendsItem = menu.findItem(R.id.feed_menu_item_friends);
        Drawable drawable = friendsItem.getIcon();
        if (drawable != null) {
            drawable.mutate();
            drawable.setColorFilter(Color.WHITE, PorterDuff.Mode.SRC_ATOP);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.feed_menu_item_friends:
                if (getActivity() != null)
                    ((MainActivity)getActivity()).showFriends();

                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void setupViewPager(View rootView) {
        viewPager = rootView.findViewById(R.id.feedViewPagerFragment);
        viewPager.setOffscreenPageLimit(3);

        FeedViewPagerAdapter viewPagerAdapter = new FeedViewPagerAdapter(getChildFragmentManager());
        viewPager.setAdapter(viewPagerAdapter);

        viewPager.setCurrentItem(0);

        if (tabLayout != null) tabLayout.setupWithViewPager(viewPager);
    }

    @Override
	public void setupTabLayout(TabLayout tabLayout) {
        this.tabLayout = tabLayout;

        tabLayout.removeAllTabs();
        tabLayout.setVisibility(View.VISIBLE);
        tabLayout.addOnTabSelectedListener(onTabSelectedListener);

        if (viewPager != null) tabLayout.setupWithViewPager(viewPager);
    }

    @Override
    public void onPause() {
        super.onPause();

        if (tabLayout != null) tabLayout.removeOnTabSelectedListener(onTabSelectedListener);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        if (tabLayout != null) tabLayout.removeOnTabSelectedListener(onTabSelectedListener);
    }

    private final TabLayout.OnTabSelectedListener onTabSelectedListener = new TabLayout.OnTabSelectedListener() {
        @Override
        public void onTabSelected(TabLayout.Tab tab) {
            viewPager.setCurrentItem(tab.getPosition());
        }

        @Override
        public void onTabUnselected(TabLayout.Tab tab) { }

        @Override
        public void onTabReselected(TabLayout.Tab tab) { }
    };
}

File: isegoria/src/main/java/com/eulersbridge/isegoria/feed/FeedFragment.kt
package com.eulersbridge.isegoria.feed

import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.graphics.Color
import android.graphics.PorterDuff
import android.os.Bundle
import android.support.design.widget.TabLayout
import android.support.v4.app.Fragment
import android.support.v4.content.ContextCompat
import android.view.*
import com.eulersbridge.isegoria.*
import com.eulersbridge.isegoria.util.ui.TitledFragment
import kotlinx.android.synthetic.main.feed_fragment.*


class FeedFragment : Fragment(), TitledFragment, MainActivity.TabbedFragment {

    private var tabLayout: TabLayout? = null

    private val viewModel: FeedViewModel by lazy {
        ViewModelProviders.of(activity!!).get(FeedViewModel::class.java)
    }

    private val onTabSelectedListener = object : TabLayout.OnTabSelectedListener {
        override fun onTabSelected(tab: TabLayout.Tab) {
            viewPager?.currentItem = tab.position
        }

        override fun onTabUnselected(tab: TabLayout.Tab) {}
        override fun onTabReselected(tab: TabLayout.Tab) {}
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.feed_fragment, container, false)

        activity?.apply {
            val colour = ContextCompat.getColor(this, R.color.darkBlue)
            statusBarColour = colour
            setMultitaskColour(colour)

            // Ensure options menu from another fragment is not carried over
            invalidateOptionsMenu()

            // Hide keyboard if navigating from login
            keyboardVisible = false
        }

        setHasOptionsMenu(true)

        return rootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        setupViewPager()
    }

    override fun getTitle(context: Context?) = context?.getString(R.string.section_title_feed)

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater!!.inflate(R.menu.feed, menu)
        super.onCreateOptionsMenu(menu, inflater)

        val friendsItem = menu!!.findItem(R.id.feed_menu_item_friends)
        friendsItem.icon?.let {
            it.mutate()
            it.setColorFilter(Color.WHITE, PorterDuff.Mode.SRC_ATOP)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        return when (item!!.itemId) {
            R.id.feed_menu_item_friends -> {
                viewModel.showFriends()
                true
            }

            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun setupViewPager() {
        viewPager.apply {
            adapter = FeedViewPagerAdapter(childFragmentManager)
            offscreenPageLimit = 3
            currentItem = 0
        }

        tabLayout?.setupWithViewPager(viewPager)
    }

    override fun setupTabLayout(tabLayout: TabLayout) {
        this.tabLayout = tabLayout

        tabLayout.apply {
            tabLayout.removeAllTabs()
            tabLayout.visibility = View.VISIBLE
            tabLayout.addOnTabSelectedListener(onTabSelectedListener)

            if (viewPager != null)
                setupWithViewPager(viewPager)
        }
    }

    override fun onPause() {
        super.onPause()

        tabLayout?.removeOnTabSelectedListener(onTabSelectedListener)
    }

    override fun onDestroy() {
        super.onDestroy()

        tabLayout?.removeOnTabSelectedListener(onTabSelectedListener)
    }
}